Userland/Libraries/LibVirtGPU/Device.h

```diff
@@ -0,0 +1,90 @@
+    virtual void blit_from_depth_buffer(NonnullRefPtr<GPU::Image>, u32 level, Vector2<u32> input_size, Vector2<i32> input_offset, Vector3<i32> output_offset) override;
+    virtual void blit_to_color_buffer_at_raster_position(void const*, GPU::ImageDataLayout const&) override;
+    virtual void blit_to_depth_buffer_at_raster_position(void const*, GPU::ImageDataLayout const&) override;
+    virtual void set_options(GPU::RasterizerOptions const&) override;
+    virtual void set_light_model_params(GPU::LightModelParameters const&) override;
+    virtual GPU::RasterizerOptions options() const override;
+    virtual GPU::LightModelParameters light_model() const override;
+
+    virtual NonnullRefPtr<GPU::Image> create_image(GPU::PixelFormat const&, u32 width, u32 height, u32 depth, u32 max_levels) override;
+    virtual ErrorOr<NonnullRefPtr<GPU::Shader>> create_shader(GPU::IR::Shader const&) override;
+
+    virtual void set_sampler_config(unsigned, GPU::SamplerConfig const&) override;
+    virtual void set_light_state(unsigned, GPU::Light const&) override;
+    virtual void set_material_state(GPU::Face, GPU::Material const&) override;
+    virtual void set_stencil_configuration(GPU::Face, GPU::StencilConfiguration const&) override;
+    virtual void set_texture_unit_configuration(GPU::TextureUnitIndex, GPU::TextureUnitConfiguration const&) override;
+    virtual void set_clip_planes(Vector<FloatVector4> const&) override;
+
+    virtual GPU::RasterPosition raster_position() const override;
+    virtual void set_raster_position(GPU::RasterPosition const& raster_position) override;
+    virtual void set_raster_position(FloatVector4 const& position, FloatMatrix4x4 const& model_view_transform, FloatMatrix4x4 const& projection_transform) override;
+
+    virtual void bind_fragment_shader(RefPtr<GPU::Shader>) override;
+
+private:
+    void encode_constant_buffer(Gfx::FloatMatrix4x4 const&, Vector<float>&);
+    Protocol::ObjectHandle allocate_handle();
+    Protocol::ResourceID create_virgl_resource(VirGL3DResourceSpec&);
+    void upload_command_buffer(Vector<u32> const&);
+
+    int m_gpu_file_descriptor { 0 };
+    Protocol::ResourceID m_vbo_resource_id { 0 };
+    Protocol::ResourceID m_drawtarget { 0 };
+    Protocol::ResourceID m_depthbuffer_surface { 0 };
+    Protocol::ObjectHandle m_blend_handle { 0 };
+    Protocol::ObjectHandle m_drawtarget_surface_handle { 0 };
+    Protocol::ObjectHandle m_depthbuffer_surface_handle { 0 };
+    Protocol::ObjectHandle m_ve_handle { 0 };
+    Protocol::ObjectHandle m_frag_shader_handle { 0 };
+    Protocol::ObjectHandle m_vert_shader_handle { 0 };
+    Protocol::ObjectHandle m_rasterizer_handle { 0 };
+    Protocol::ObjectHandle m_dsa_handle { 0 };
+
+    struct VertexData {
+        float r;
+        float g;
+        float b;
+        float x;
+        float y;
+        float z;
+    };
+    Vector<VertexData> m_vertices;
+
+    Vector<float> m_constant_buffer_data;
```

Comment:
`Vector<float>` can be replaced with `FloatVector` everywhere (similar to `FloatMatrix4x4` instead of `Matrix4x4<float>`, which was already done in this PR)

Reply:
This is actually a `AK::Vector` of floats though :)

Reply:
Ah that makes a lot of sense. Apologies, this can be resolved!